doc/source/raysgd/v2/migration-guide.rst

```diff
@@ -0,0 +1,268 @@
+
+There are 3 ways to get state during or after the training execution:
+
+
+#. Return values from your training function
+#. Intermediate results via ``sgd.report()``
+#. Saving & loading checkpoints via ``sgd.save_checkpoint()`` and ``sgd.load_checkpoint()``
+
+Return Values
+~~~~~~~~~~~~~
+
+To get any state from training *after* training has completed, you can simply return it from your training function. The return values from each the workers will be added to a list and returned from the ``trainer.run()`` call.
+
+.. code-block:: python
+
+   def train_func():
+       model = Net()
+       trainer_loader = MyDataset()
+       for batch in train_loader:
+           model.train(batch)
+
+       return model
+
+   trainer = Trainer(backend="torch")
+   trainer.start()
+   results = trainer.run(train_func, num_workers=2)
+   assert len(results) == 2
+   trained_model = results[0]
+
+Intermediate Reporting
+~~~~~~~~~~~~~~~~~~~~~~
+
+If you want to access any values *during* the training process, you can do so via ``sgd.report()``. You can pass in any values to ``sgd.report()`` and these values from all workers will be sent to any callbacks passed into your ``Trainer``.
+
+.. code-block:: python
+
+   from ray import sgd
+   from ray.sgd import SGDCallback, Trainer
+   from typing import List, Dict
+
+   class PrintingCallback(SGDCallback):
+       def handle_result(self, results: List[Dict], **info):
+           print(results)
+
+   def train_func():
+       for i in range(3):
+           sgd.report(epoch=i)
+
+   trainer = Trainer(backend="torch", num_workers=2)
+   trainer.start()
+   result = trainer.run(
+       train_func,
+       callbacks=[PrintingCallback()]
+   )
+   # [{'epoch': 0, '_timestamp': 1630471763, '_time_this_iter_s': 0.0020279884338378906, '_training_iteration': 1}, {'epoch': 0, '_timestamp': 1630471763, '_time_this_iter_s': 0.0014922618865966797, '_training_iteration': 1}]
+   # [{'epoch': 1, '_timestamp': 1630471763, '_time_this_iter_s': 0.0008401870727539062, '_training_iteration': 2}, {'epoch': 1, '_timestamp': 1630471763, '_time_this_iter_s': 0.0007486343383789062, '_training_iteration': 2}]
+   # [{'epoch': 2, '_timestamp': 1630471763, '_time_this_iter_s': 0.0014500617980957031, '_training_iteration': 3}, {'epoch': 2, '_timestamp': 1630471763, '_time_this_iter_s': 0.0015292167663574219, '_training_iteration': 3}]
+   trainer.shutdown()
+
+See the :ref:`v2 User Guide <sgd-user-guide>` for more details.
+
+Checkpointing
+~~~~~~~~~~~~~
+
+Finally, you can also use ``sgd.save_checkpoint()`` and ``sgd.load_checkpoint()`` to write checkpoints to disk during the training process, and to load from the most recently saved checkpoint in the case of node failures.
+
+See the :ref:`Checkpointing <sgd-checkpointing>` and :ref:`Fault Tolerance & Elastic Training <sgd-fault-tolerance>` sections on the user guide for more info.
+
+.. _sgd-migration-tune:
+
+Hyperparameter Tuning with Ray Tune
```

Comment:
Mention SGD v1 API and also link the Tune Integration section from V2 docs?

Reply:
Done